docs(HIP-632): Decouple from virtual addresses; referring only to aliases (#1011)
hashgraph/hedera-improvement-proposal: HIP/hip-632.md

last-call-date-time: 2023-01-10T07:00:00Z
created: 2022-11-28
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/638
updated: 2024-07-15
requires: 583
---

## Abstract

> Hedera accounts have the capability to have a linked _EVM address alias_, an EVM address.  In this way smart contracts can deal

Reviewer: perhaps it'll be helpful to link/refer to the table in HIP 583 that defines the various address formats.

## User stories

1. As a developer I want to be able to distinguish between Hedera Account ID and aliases within a smart contract so I can resolve address mappings in balance, permission or similar scenarios.
2. As a developer I want to be able to confirm account authorization within a contract using simple `ECDSA`/`ED2519` key raw signatures as I would on Ethereum using `ECRECOVER`.
3. As a developer I want to be able to confirm account authorization for `ECDSA` accounts within a contract without being limited to `ECRECOVER`.
4. As a developer I want to be able to confirm account authorization for accounts within a contract using protobuf signature maps in a similar fashion as I would on Ethereum using `ECRECOVER`.
  
## Specification

A new system contract with functions to support address (Hedera Account ID and EVM address alias) mapping and authorization is proposed.

This will aid developers to support smart contract migration logic for cases where aliases are encountered.
Additionally, EVM developers, who were limited to `ECRECOVER` authorization flows, will be able to expand
authorization checks to other key types. 

> HAS methods here can use EVM-styled 20-byte addresses which are either Hedera account num aliases or EVM address 

Reviewer: Perhaps define `HAS` before it is used here?
```suggestion
HAS methods can use EVM-styled 20-byte addresses which are either Hedera account num aliases or EVM address 
```

aliases. 

* In this document the word _alias_, by itself, means either kind of alias.

  (n.b.: See HIP-583 for definitions of Hedera account num alias and EVM address alias.  Hedera key aliases, defined there,
are not supported by these methods.)

| hash | signature | return | description |
| --- | --- | --- | --- |

> | `0xdea3d081` | getEvmAddressAlias(address) | address | given a Hedera Account ID (account num alias) returns the EVM address alias, if present |

Reviewer: started implementation and noticed that the return type should probably be `(int64 responseCode, address alias)` to be consistent with the other view functions.  The same for `getHederaAccountNumAlias` below

Author: done

last-call-date-time: 2023-01-10T07:00:00Z
created: 2022-11-28
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/638
updated: 2024-09-10
requires: 583
---

## Abstract

Hedera accounts have the capability to have a linked _EVM address alias_, an EVM address.  In this way smart contracts can deal
interact with Hedera accounts.  

Accounts now have an organized presentation of different address formats that support historical formats and EVM equivalent
formats.  These addresses may be stored within smart contract state but their usefulness is limited without the ability to differentiate addresses.

These methods will be placed in the system contract for the _Hedera Account Service_ (`HAS`). 

## Motivation

Since the same account may be referred to with two different 20-byte identifiers (EVM address alias or Hedera account num alias),
to avoid issues of loss of balance and permissions it is important that developers can distinguish between Hedera Account IDs 
and EVM addresses and translate easily between them.

Many smart contracts utilize `ECRECOVER` as part of their authorization process in smart contracts.

> However, this is limited to `ECDSA` based signatures, which leaves accounts that have `E25519` and complex key types, supported by Hedera,

Reviewer 2: ```suggestion
However, this is limited to `ECDSA` based signatures, which leaves accounts that have `ED25519` and complex key types, supported by Hedera,
```

last-call-date-time: 2023-01-10T07:00:00Z
created: 2022-11-28
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/638
updated: 2024-09-13
requested-by: David Bakin <@david-bakin-sl>
requires: 583
---

## Abstract

> Hedera accounts have the capability to have a linked _EVM address alias_, an EVM address.  In this way smart contracts can deal

Reviewer 3: Maybe?
```suggestion
Hedera accounts have the capability to have a linked _EVM address alias_, an EVM address.  In this way the EVM can represent these accounts in an industry standard format
```

  
## Specification

A new system contract with functions to support address (Hedera Account ID and EVM address alias) mapping and authorization is proposed.

This will aid developers to support smart contract migration logic for cases where aliases are encountered.
Additionally, EVM developers, who were limited to `ECRECOVER` authorization flows, will be able to expand
authorization checks to other key types. 

> HAS methods here can use EVM 20-byte addresses which are either Hedera account num aliases or EVM address 

Reviewer 3: ```suggestion
HAS methods here can use EVM styled 20-byte addresses which are either Hedera account num aliases or EVM address 
```

aliases. 

* In this document the word _alias_, by itself, means either kind of alias.

  (n.b.: See HIP-583 for definitions of Hedera account num alias and EVM address alias.  Hedera key aliases, defined there,
are not supported by these methods.)

| hash | signature | return | description |
| --- | --- | --- | --- |
| `0xdea3d081` | getEvmAddressAlias(address) | address | given a Hedera Account ID (account num alias) returns the EVM address alias, if present |
| `0xbbf12d2e` | getHederaAccountNumAlias(address) | address | given an alias returns the top level Hedera Account ID (as account num alias), if applicable |

> | `0xcd82a665` | isAlias(address) | bool | true if an account num alias or evm address alias), false if non existing account |

Reviewer 3: I think this should be isEvmAddressAlias()
```suggestion
| `0xcd82a665` | isEvmAddressAlias(address) | bool | true if a valid evm address alias on the network, false if an account num alias or non existing account |
```

Author: What we had was `isVirtualAddress` defined as `true if valid virtual address, false if long-zero or non existing account`.  

But we want to do two things at once:  Change the wording `virtual address` to `evm address alias` **and** remove the limitation "can't be a long-zero" which we felt was not a restriction we needed (or that I could find any reason for among us).  

So what we're left with, if we want to go that route, is "is evm address alias or long-zero _== account num alias_ and it must be for an existing account", i.e., is an alias and the account it is an alias for exists.  (Even hollow accounts.). Thus the name.

Do we still want to go in that direction (i.e., _also_ eliminate the restriction that it can't be a long-zero)?

Perhaps call it `isValidAlias` to make it explicit that it must be for a valid (== existing) account.

         * ECDSA p-384 signature
         */
        bytes ECDSA_384 = 5;

> gi tpush

Reviewer 2: Probably an unintentional addition?

Author: oops!

## Reference Implementation

## (New) Response Codes That Can Be Returned

> * `INVALID_ARGUMENT_TO_SYSTEM_CONTRACT_METHOD` - an argument to a system contract method is invalid (e.g., byte array of wrong length)

Reviewer: Have we checked that there are not existing response codes that could be used for these purposes.  Just wondering as adding response codes I believe at entails updating the HederaResponseCodes.sol file at the very least.

Author: I did check, didn't find anything that looked appropriate.
